core/math/random.cpp

```diff
@@ -158,20 +158,24 @@ Array Random::choices(const Variant &p_from, int p_count, const PoolIntArray &p_
 		}
 	}
 
+	ERR_FAIL_COND_V_MSG((weights.empty() && p_is_cumulative), Array(), "Cumulative weights cannot be empty.");
 	if (!weights.empty()) {
 		if (p_is_cumulative) {
+			int prev_cumulative = weights[0];
+			for(int i = 1; i < weights.size(); i++) {
+				ERR_FAIL_COND_V_MSG((weights[i] < prev_cumulative), Array(), "Cumulative weights must be non-decreasing.");
+				prev_cumulative = weights[i];
+			}
 			sum = weights[weights.size() - 1];
 			cumulative_weights = weights;
 		} else {
 			for (int i = 0; i < weights.size(); i++) {
-				if (weights[i] < 0) {
-					ERR_FAIL_V_MSG(Array(), "Weights must be positive integers.");
-				} else {
-					sum += weights[i];
-					cumulative_weights.push_back(sum);
-				}
+				ERR_FAIL_COND_V_MSG((weights[i] < 0), Array(), "Weights must be non-negative integers.");
+				sum += weights[i];
+				cumulative_weights.push_back(sum);
 			}
 		}
+		ERR_FAIL_COND_V_MSG((sum == 0), Array(), "Sum of weights cannot be zero");
```

Comment:
Note that wrapping an expression is not required in error macros:
```suggestion
		ERR_FAIL_COND_V_MSG(sum == 0, Array(), "Sum of weights cannot be zero");
```
This is not required to fix, but just for the future.